tracer/src/Datadog.Trace/AppSec/Coordinator/SecurityCoordinator.Framework.cs

```diff
@@ -249,22 +249,26 @@ private SecurityCoordinator(Security security, Span span, HttpTransport transpor
     [MethodImpl(MethodImplOptions.NoInlining)]
     private static bool TryGetUsingIntegratedPipelineBool() => HttpRuntime.UsingIntegratedPipeline;
 
-    internal Dictionary<string, object> GetBodyFromRequest()
+    internal Dictionary<string, object>? GetBodyFromRequest()
     {
-        var formData = new Dictionary<string, object>(_httpTransport.Context.Request.Form.Keys.Count);
-        foreach (string key in _httpTransport.Context.Request.Form.Keys)
+        var form = RequestDataHelper.GetForm(_httpTransport.Context.Request);
+
+        if (form is null)
+        {
+            return null;
+        }
+
+        var formData = new Dictionary<string, object>(form.Keys.Count);
```

Comment:
I wonder, does the above protection _definitely_ protect you? 🤔 i.e. this code suggests you get an exception when, you first access `_httpTransport.Context.Request.Form` (hence the protection above)?

If so, then it seems _unlikely_ that we need the protections below any more? And if we _do_ then it seems _very_ likely that `form.Keys` would _also_ throw, no 🤔 


I think it would be nice if we had some tests (integration or unit) that replicates the verification that aspnet does so we can know we're actually protecting ourselves correctly 🙂

Reply:
I was basically checking the source code to find places where the exceptions could potentially be thrown.

HttpRequest.Form (Get) calls ValidateHttpValueCollection. Same for GetValues and Get. Keys does not seem to check that.

I tried to reproduce the case where getting the form does not throw exception but later, getting the values does but did not get it. But the stack in some errors detected in customers shows that this can happen. Anyway, I will add some unit tests trying to cover all the cases.

Reply:
I have added some tests. Basically, when normal validation is enabled, accessing request.Form would trigger the exception.
HttpRequest also allows to perform what they call "granular validation", which is different than the normal mode. This mode translates validation to the HttpCollections fields of the request by adding a callback to each of them that is called when their values are accessed, instead of being done at a httpRequest level when accessing the property Form.
This mode seems to be enabled by the framework and I could not find an easy way to enable it without reflection. The involved methods are internal and sometimes use classes that cannot be accessed out of the scope of the assembly without reflection.
In summary: a query with regular validation will throw an exception when accessing Form. A query with granular validation would not throw an exception when accessing Form but it will throw it when accessing GetValues or Get. The keys property never throws any exception.
The granular validation mode would explain why we have some exceptions that where thrown when calling GetValues but did not thrown it earlier, when accessing Form.

[https://app.datadoghq.com/error-tracking?query=service%3Ainstrumentation-telemetry-data%20%2Aappsec%2A%20%40lib_language%3Adotnet&fromUser=false&issueId=061ff9ca-0ddb-11ef-bfc4-da7ad0900002&refresh_mode=sliding&source=all&from_ts=1730449059747&to_ts=1731053859747&live=true](https://app.datadoghq.com/error-tracking?query=service%3Ainstrumentation-telemetry-data%20%2Aappsec%2A%20%40lib_language%3Adotnet&fromUser=false&issueId=061ff9ca-0ddb-11ef-bfc4-da7ad0900002&refresh_mode=sliding&source=all&from_ts=1730449059747&to_ts=1731053859747&live=true)

Reply:
:TIL: Interesting, thanks for the explanation! It might be worth adding some of that to the code itself 🙂 Thanks!

Reply:
Done! Thanks!